quinn-udp/src/windows.rs

```diff
@@ -76,30 +114,100 @@ impl UdpSocketState {
+    pub fn send(&self, socket: UdpSockRef<'_>, transmits: &[Transmit]) -> io::Result<usize> {
         let mut sent = 0;
         for transmit in transmits {
-            match socket.0.send_to(
-                &transmit.contents,
-                &socket2::SockAddr::from(transmit.destination),
-            ) {
-                Ok(_) => {
-                    sent += 1;
+            // we cannot use [`socket2::sendmsg()`] and [`socket2::MsgHdr`] as we do not have access
+            // to the inner field which holds the WSAMSG
+            let mut ctrl_buf = cmsg::Aligned([0; CMSG_LEN]);
+            let daddr = socket2::SockAddr::from(transmit.destination);
+
+            let mut data = WinSock::WSABUF {
+                buf: transmit.contents.as_ptr() as *mut _,
+                len: transmit.contents.len() as _,
+            };
+
+            let ctrl = WinSock::WSABUF {
+                buf: ctrl_buf.0.as_mut_ptr(),
+                len: ctrl_buf.0.len() as _,
+            };
+
+            let mut wsa_msg = WinSock::WSAMSG {
+                name: daddr.as_ptr() as *mut _,
+                namelen: daddr.len(),
+                lpBuffers: &mut data,
+                Control: ctrl,
+                dwBufferCount: 1,
+                dwFlags: 0,
+            };
+
+            // Add control messages (ECN and PKTINFO)
+            let mut encoder = unsafe { cmsg::Encoder::new(&mut wsa_msg) };
+
+            if let Some(ip) = transmit.src_ip {
+                let ip = std::net::SocketAddr::new(ip, 0);
+                let ip = socket2::SockAddr::from(ip);
+                match ip.family() {
+                    WinSock::AF_INET => {
+                        let src_ip =
+                            unsafe { ptr::read(ip.as_ptr() as *const WinSock::SOCKADDR_IN) };
+                        let pktinfo = WinSock::IN_PKTINFO {
+                            ipi_addr: src_ip.sin_addr,
+                            ipi_ifindex: 0,
+                        };
+                        encoder.push(WinSock::IPPROTO_IP, WinSock::IP_PKTINFO, pktinfo);
+                    }
+                    WinSock::AF_INET6 => {
+                        let src_ip =
+                            unsafe { ptr::read(ip.as_ptr() as *const WinSock::SOCKADDR_IN6) };
+                        let pktinfo = WinSock::IN6_PKTINFO {
+                            ipi6_addr: src_ip.sin6_addr,
+                            ipi6_ifindex: unsafe { src_ip.Anonymous.sin6_scope_id },
+                        };
+                        encoder.push(WinSock::IPPROTO_IPV6, WinSock::IPV6_PKTINFO, pktinfo);
+                    }
+                    _ => {
+                        return Err(io::Error::from(io::ErrorKind::InvalidInput));
+                    }
                 }
+            }
+
+            // ECN is a C integer https://learn.microsoft.com/en-us/windows/win32/winsock/winsock-ecn
+            let ecn = transmit.ecn.map_or(0, |x| x as i32);
```

Comment:
Shouldn't we use the C integer type, then?
```suggestion
            // ECN is a C integer https://learn.microsoft.com/en-us/windows/win32/winsock/winsock-ecn
            let ecn = transmit.ecn.map_or(0, |x| x as libc::c_int);
```